packages/docusaurus-plugin-search/theme/useSearch.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {useContext} from 'react';
+import {groupBy, sortBy} from 'lodash-es';
+import {useBaseUrlUtils} from '@docusaurus/useBaseUrl';
+import Fuse from 'fuse.js';
+
+const ELEMENTS = {
+  h1: 3,
+  h2: 2,
+  h3: 1,
+  p: 0.5,
+};
+
+const score = (elements) =>
+  elements.reduce((acc, cur) => acc + (1 - cur.score) * ELEMENTS[cur.type], 0);
+
+const sortType = (data) => sortBy(data, 'type');
+
+const rank = (list) => {
+  const result = list.slice(0, 100).filter((v) => v.score < 0.5);
+  const data = groupBy(
+    result.map((v) => ({...v.item, match: v.matches})),
+    'pageID',
+  );
+  const ranked = sortBy(
+    Object.entries(data).map(([key, value]) => ({
+      key,
+      value: score(value),
+    })),
+    'value',
+  ).reverse();
+  return ranked.map(({key}) => ({key, value: sortType(data[key])}));
+};
+
+export const FuseContext = React.createContext(undefined);
+
+export function useSearch() {
+  const {withBaseUrl} = useBaseUrlUtils();
+  const {fuse, setFuse} = useContext(FuseContext);
+  const getData = React.useCallback(() => {
+    return Promise.all([
+      fetch(withBaseUrl('search_index.json')).then((res) => res.json()),
+      fetch(withBaseUrl('search_result.json')).then((res) => res.json()),
+    ]).then((res) => {
+      const newFuse = new Fuse(
+        res[1],
+        {
+          threshold: 1,
+          distance: 100,
+          minMatchCharLength: 3,
+          includeMatches: true,
+          includeScore: true,
+        },
+        Fuse.parseIndex(res[0]),
+      );
+      setFuse(newFuse);
+      return newFuse;
+    });
+  }, [setFuse]);
+  return async (term) => {
+    if (!fuse) {
+      const newFuse = await getData();
+      return rank(newFuse.search(term));
+    }
+    return rank(fuse.search(term));
+  };
+}
```

Comment:
All this looks not ideal to me.

We need a way to "preload" the search engine as soon as the user hover the search bar, even before the first keystroke.

We should only fetch the data once, and initialize FuseJS once, not on

We should rely on FuseJS scoring/orderi,ng

Reply:
Oh didn't see but you set the fuse engine in context so it does not re-execute, not fan of how it's done but. seems to work not too bad 👍